implement iscsi persistent disk volume plugin

rootfs opened the pull request:
This pull request implemented an iSCSI disk volume plugin. The idea is to allow on-premise kube setup to persist data in a block volume through iSCSI channel and help DevOps run test in house even without access to GCE.

The code is heavily inspired by GCE Persistent Disk.

A sample pod is provided. This pod runs on a Fedora 21 based node and persists to iSCSI backed volume.

Background information, use cases,  high level design, and how-to can be found [here](http://rootfs.github.io/iSCSI-Kubernetes/)

@thockin @markturansky @smarterclayton @eparis

markturansky commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,167 @@
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume/gce_pd"
```
I don't think details from one plugin should leak into another.

If there are generic mounting utils in GCE that other volumes would benefit from, it is probably better to refactor those commons things into a package all volumes can use.

rootfs replied:
add @thockin, i think we probably have some code to refactor and share.

markturansky commented:
@calfonso

calfonso commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,167 @@
+		return errors.New("path busy")
+	}
+	//glog.Infof("iSCSIPersistentDisk detach disk:  umount:%s", globalPDPath)
+	if err := iscsi.mounter.Unmount(globalPDPath, 0); err != nil {
```
Is there a benefit to using the mounter.Unmount/Mount approach over using syscall.Mount/Unmount?

rootfs replied:
mounter is an interface that helps setup a unit test. Under the hood, it calls syscall.

rootfs commented on examples/iscsi-pd/iscsi-pd.json:
```diff
@@ -0,0 +1,59 @@
+{
+    "id": "iscsipd",
+    "kind": "Pod",
+    "apiVersion": "v1beta1",
```
will do, thanks

thockin replied:
I was not aware we had started that - some of v1b3 is still changing...

markturansky replied:
I was told to use v1b3 examples.  I assumed that was broadly applied.  I can correct my examples...

markturansky replied:
Please include this in examples_test.go.

Also, all other examples are being remade into v1beta3.  Is it not time for this one, too?

thockin replied:
I see this comment is not addressed - we have started converting examples over to v1beta3...

thockin commented on pkg/api/types.go:
```diff
@@ -221,6 +224,17 @@ type GCEPersistentDisk struct {
 }
 
+// A ISCSI Disk can only be mounted as read/write once.
+type ISCSIDisk struct {
```
iSCSI is a lot less cloud-specific than GCE PD, so maybe the answer is yes, but...

Do we really want to expose this as a discrete type?  Maybe we want to glue this into @markturansky 's larger storage refactoring instead?  This is just another form of pet storage...

markturansky replied:
@thockin I understand and agree with the idea that users defining their own pets in VolumeSource will eventually go away (replacing the network volumes with a struct like you suggested in #4055).

This plugin and @calfonso NFSMount plugin are orthogonal issues that can be developed independently of what I'm doing, but then refactored into whatever design comes next.  I have no problem doing that refactoring work as a follow-on task.

rootfs replied:
Adding to that, iSCSI has some cloud affinity, it is the default implementation for OpenStack Cinder.

thockin replied:
Sure, but iSCSI and NFS could both be implemented by different kinds of backends in different environments.  while GCE PD acts like a block device it does not implement an iSCSI protocol. :)

thockin commented on pkg/kubelet/volume/gce_pd/mount_util.go:
```diff
@@ -23,7 +23,7 @@ import (
 // number of references to that device. Returns both the full device path under
 // the /dev tree and the number of references.
-func getMountRefCount(mounter mount.Interface, mountPath string) (string, int, error) {
+func GetMountRefCount(mounter mount.Interface, mountPath string) (string, int, error) {
```
This should all be extracted to pkg/util/mount or something like that

thockin commented on pkg/kubelet/volume/gce_pd/mount_util_unix.go:
```diff
@@ -26,7 +26,7 @@ import (
 // with the device for it's parent.  If they are the same, it's not a mountpoint, if they're
 // different, it is.
-func isMountPoint(file string) (bool, error) {
+func IsMountPoint(file string) (bool, error) {
```
likewise

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,252 @@
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume/gce_pd"
```
yeah, no cross-linkages, please

thockin commented:
API question aside, I have a big problem with the duplication.  I think we can implement this in 2 layers.

First is a utility lib that handles mounting and unmounting an arbitrary block device - it will need a bunch of stuff injected to parameterize the control.

Second is GCEPD and iSCSI (and eventually more) which are implemented purely in terms of the block-device layer.

That way, when we realize we've (again) done something dumb in the unmount-and-cleanup path, we fix it once.

Commit:
fix header so boilerpate can pass

Signed-off-by: Huamin Chen <hchen@redhat.com>

Commit:
- refactor kubelet volume mount util
- refactor persistent disk:
   * move pdManager to package disk
   * extract common setup and teardown logic

Signed-off-by: Huamin Chen <hchen@redhat.com>

rootfs commented:
@thockin here is a refactor:
- move mount_util from gce-pd to volume package
- move pdManager and SetUp/TearDown utility methods into disk package, and let iscsi-pd to be the first consumer of this package.

I haven't changed gce-pd yet. If these look good to you, I'll update gce-pd in another commit.

Commit:
iscsi persisten disk: reconcile with master branch changes

Commit:
Rebase to reconcile with pkg/util/mount refactor

eparis commented:
@thockin @pmorie @markturansky @calfonso 

This looks ready to merge to me.  He just rebased on top of the latest rewrite.  Please review and LGTM?

markturansky commented on examples/iscsi-pd/README.md:
```diff
@@ -0,0 +1,56 @@
+then edit */etc/iscsi/initiatorname.iscsi* and */etc/iscsi/iscsid.conf* to match your iSCSI target configuration.
+
+I mostly follow these [instructions](http://www.server-world.info/en/note?os=Fedora_21&p=iscsi&f=2) to setup iSCSI initiator and these [instructions](http://www.server-world.info/en/note?os=Fedora_21&p=iscsi) to setup iSCSI target.
+
```
Could those instructions be explained succinctly in this README?  Notes in another site could 404 and having a single concise guide in Kubernetes is helpful.

pmorie replied:
Do we not have a setup guide for F21 inside the kube repo?

rootfs replied:
yes, i followed the guide. But instructions here are about iSCSI target/initiator setup.

rootfs replied:
@markturansky will the README be part of the site?

markturansky replied:
The README will live in examples/iscsi-pd and be the canonical guide for mounting iSCSCI disks.

rootfs replied:
couple of README in other examples look even lengthier. Would like to know if there is a guideline.

markturansky replied:
I don't know of a guideline or standard to follow.

markturansky commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+func (plugin *ISCSIDiskPlugin) Name() string {
+	if plugin.legacyMode {
```
What is legacyMode for this new volume?   I see this in gce_pd, but I assume this is because that implementation is one of the earlier volumes.  I assume there are no legacy iSCSCI disks to handle.  

That's two assumptions that might bite me in the rear.

rootfs replied:
legacyMode is in every volume plugin but is currently unused. My naive understanding is that legacyMode is really something left for the future :)

markturansky replied:
Right, because all 4 volumes existed before @thockin made the current volumes framework we're using.

I'd like to learn if new volumes need that, too.  I didn't include it in the volume I am making.  I don't know if that's right or wrong.

pmorie replied:
IMO: new plugins do not need legacy mode, since you cannot create new volumes with a legacy mode plugin.

thockin replied:
This - plugins that existed before teh new namespaced names have legacy names.  New ones should not.

thockin commented on cmd/kubelet/app/plugins.go:
```diff
@@ -41,6 +42,6 @@ func ProbeVolumePlugins() []volume.Plugin {
 	allPlugins = append(allPlugins, host_path.ProbeVolumePlugins()...)
 	allPlugins = append(allPlugins, secret.ProbeVolumePlugins()...)
-
+	allPlugins = append(allPlugins, iscsi_pd.ProbeVolumePlugins()...)
```
Is "_pd" a useful suffix?  We use it on gce_pd because that is literally the product name "Persistent Disk".  Could we just call this pkg "iscsi" ?

rootfs replied:
ok

thockin commented on pkg/api/types.go:
```diff
@@ -221,6 +224,25 @@ type GCEPersistentDiskVolumeSource struct {
+// A ISCSI Disk can only be mounted as read/write once.
+type ISCSIDiskVolumeSource struct {
+	// iSCSI target portal
+	Portal string `json:"portal,omitempty" description:"iscsi target portal"`
```
The word "portal" has specific meaning in kubernetes - can you either rename this or document that it does not necessarily mean the same thing as a kubernetes portal

rootfs replied:
will do. 

unfortunately portal also has specific meaning in iscsi

thockin replied:
Also document the expected format of this

thockin commented on pkg/api/types.go:
```diff
@@ -221,6 +224,25 @@ type GCEPersistentDiskVolumeSource struct {
+	// iSCSI target portal
+	Portal string `json:"portal,omitempty" description:"iscsi target portal"`
+	// iSCSI target iqn
+	IQN string `json:"iqn,omitempty" description:"iscsi target iqn"`
```
Please expand acronyms in comments, for teh sake of those of use who don't know what it means

thockin commented on pkg/api/types.go:
```diff
@@ -180,6 +180,9 @@ type VolumeSource struct {
 	Secret *SecretVolumeSource `json:"secret"`
+	// ISCSIDiskVolumeSource represents an ISCSI Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	ISCSIDisk *ISCSIDiskVolumeSource `json:"iscsiDisk" description:"iSCSI disk resource attached to the host machine on demand"`
```
I think we're NOT doing description tags in the internal API file

thockin commented on pkg/api/types.go:
```diff
@@ -221,6 +224,25 @@ type GCEPersistentDiskVolumeSource struct {
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs"
+	// TODO: how do we prevent errors in the filesystem from compromising the machine
+	// TODO: why omitempty if required?
```
if it is required, don't omitempty.  Is it required or can it be "figured out" ?

thockin commented on pkg/api/types.go:
```diff
@@ -221,6 +224,25 @@ type GCEPersistentDiskVolumeSource struct {
+	// iSCSI target iqn
+	IQN string `json:"iqn,omitempty" description:"iscsi target iqn"`
+	// iSCSI target lun number
+	Lun int `json:"lun,omitempty" description:"iscsi target lun number"`
```
please comment required-ness.  don't omitempty for required fields.

thockin commented on pkg/kubelet/pd/doc.go:
```diff
@@ -0,0 +1,19 @@
+
+// Package disk includes internal representations of external persistent
+// as well as utility methods required to setup and tear down disk volume
+package disk
```
The pkg and the dir should be named the same unless real circumstances

thockin commented on pkg/kubelet/pd/disk.go:
```diff
@@ -0,0 +1,97 @@
+
+// Abstract interface to PD operations.
+type PDManager interface {
+	MakeGlobalPDName(disk interface{}) string
```
Needs a comment

thockin commented on pkg/kubelet/pd/disk.go:
```diff
@@ -0,0 +1,97 @@
+	"github.com/golang/glog"
+)
+
+// Abstract interface to PD operations.
```
This could detail a bit more about what this interface is about.  Given that this is to be a common interface, please err on the side of too many comments.

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.Plugin {
+	return []volume.Plugin{&ISCSIDiskPlugin{nil, false, "", "", ""},
```
if you're going to line-wrap wrap some of the slice elements, please wrap them all

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+
+type ISCSIDiskPlugin struct {
+	host       volume.Host
+	legacyMode bool // if set, plugin answers to the legacy name
```
You don't have a legacy name - you're a new plugin.  You can drop all the legacy junk

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+type ISCSIDiskPlugin struct {
+	host       volume.Host
+	legacyMode bool // if set, plugin answers to the legacy name
+	portal     string
```
Why does the plugin have these properties?  The Plugin object represents the way to create instances, but not any particular instance.

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+	if spec.ISCSIDisk.Lun != 0 {
+		lun = strconv.Itoa(spec.ISCSIDisk.Lun)
+	}
+	plugin.portal = spec.ISCSIDisk.Portal
```
same as previous comment - the plugin should not have any dynamic properties , those are for instances.

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+		manager:    manager,
+		mounter:    mounter,
+		plugin:     plugin,
+		portal:     plugin.portal,
```
Ahh, I see what you're doing.  NAK.  Cleaner has to handle the case of kubelet crashing, and when it recovers it finds a volume directory that it has no Pod for.  any state you need in Cleaner MUST be recoverable from the system.  Either use the name or dir to look up the mount info in /proc/mounts or else embed info into the name you write to disk.

For example, given an iSCSI with IQN of "x" and LUN of "y" you might make the global mountpoint be iscsi/x_y - then you could parse out IQN and LUN later.

But do you really need these fields for teardown?

rootfs replied:
for tear down, i need to know the portal (i.e. iscsi target ip), so i can tell iscsi initiator to logout.

thockin replied:
then that needs to be recoverable from on-disk state.

Consider the case of:

1) API server has a pod with an iscsi
2) kubelet crashes
3) APi server removes pod with iscsi
4) kubelet recovers - it finds a volume dir, but has no pod state

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_pd.go:
```diff
@@ -0,0 +1,189 @@
+	globalPDPath := iscsi.manager.MakeGlobalPDName(*iscsi)
+	// make mountpoint rw/ro work as expected
+	//FIXME revisit pkg/util/mount and ensure rw/ro is implemented as expected
+	if iscsi.readOnly {
```
Can you capture this as

````
mode = "rw"
if readonly {
    mode = "ro"
}
exec(mount, "remount,"+mode, ...)

rootfs replied:
will do

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,127 @@
+
+func probeDevicePath(devicePath string, maxRetries int) bool {
+	numTries := 0
+	for {
```
why not write this as `for i := 0; i < maxretries; i++ {` ? much less scary

rootfs replied:
will do

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,127 @@
+	return false
+}
+
+func makePDNameInternal(host volume.Host, portal string, iqn string, lun string) string {
```
can you write a comment about the structure of the name

thockin replied:
e.g. why you use multiple levels of path - I am not sure teh cleanup & recovery path will handle this properly..

rootfs replied:
will do

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,127 @@
+
+type ISCSIDiskUtil struct{}
+
+func (util *ISCSIDiskUtil) MakeGlobalPDName(disk interface{}) string {
```
Ugh, is there no way to factor this without bouncing through interface{} ?

rootfs replied:
not to my awareness :)

thockin replied:
I have a proposal to make this an easier review.

First offer a PR that makes this transformation on GCE PD so we can evaluate things like interface{} - I bet there's a cleaner factoring.

Second, implement iscsi in those terms

rootfs replied:
ok, let's do this way. any contact you know has ideas other than interface{}?

thockin replied:
I think the problem is that you are trying to go from type-specific code to type-agnostic code to type-specific code.  I think there might be a different way to slice it, but I need to look at it more.  I can probably live with interface{} right now, but I'd like to see how it changes the PD case.

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,127 @@
+	exist := probeDevicePath(devicePath, 1)
+	if exist == false {
+		// discover iscsi target
+		_, err := iscsi.execCommand("iscsiadm", []string{"-m", "discovery", "-t", "sendtargets", "-p",
```
maybe we should check for the existence of iscsiadm in the Plugin CanSupport() function, and log something to the effect of "I would handle this but I don't have iscsiadm"

rootfs replied:
if no iscsiadm, will exec return error? will add a check for that.

thockin replied:
I'm 99% sure it will return an error - if iscsiadm has a -v flag or something that is ideal.

thockin commented on pkg/kubelet/volume/iscsi_pd/iscsi_util.go:
```diff
@@ -0,0 +1,127 @@
+			return err
+		}
+		// login to iscsi target
+		_, err = iscsi.execCommand("iscsiadm", []string{"-m", "node", "-p",
```
don't line-wrap in arbitrary places - I'd rather have one long line to read

thockin commented on pkg/util/mount/mount.go:
```diff
@@ -76,3 +80,40 @@ func GetMountRefs(mounter Interface, mountPath string) ([]string, error) {
+	return refCount, nil
+}
+
+// given a device path, find the mount on that device from /proc/mounts
```
Not godoc compatible (should start with function name).  Please document the return values.

thockin commented on pkg/util/mount/mount.go:
```diff
@@ -76,3 +80,40 @@ func GetMountRefs(mounter Interface, mountPath string) ([]string, error) {
 }
+
+// given a device path, find its reference count from /proc/mounts
+func GetDeviceRefCount(mounter Interface, deviceName string) (int, error) {
```
Not godoc compatible (should start with function name).

rootfs replied:
will do, and any hook can detect it?

thockin replied:
no, it's just general feedback - public functions should have godoc-compatible comments

rootfs replied:
nice to know, thanks

thockin commented:
Thanks for undertaking this refactoring.  The biggest problem I have is that the refactoring is tightly coupled to the iscsi changes.  Also, we should refactor gce_pd to ensure the refactoring works.  I'm worried about the interface{} and I think we can do better, but it's hard to see with the new iscsi code mixed in.

Commit:
address review feedbacks

Signed-off-by: Huamin Chen <hchen@redhat.com>

AntonioMeireles committed:
Add Docker caching support to CoreOS' clusters.

    a more idiomatic take on #4378. as a bonus, and to be nice on
    resources, an ultra lightweight docker-registry container is
    consumed.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

AntonioMeireles committed:
simplify fleet invocation across CoreOS  cloud-configs.

Signed-off-by: António Meireles <antonio.meireles@reformi.st>

derekwaynecarr committed:
Client support to delete events

Commit:
Update to option enable profiling on the master daemon processes.
--profiling=true , default is false

brendandburns committed:
Add kubelet to the master.

Commit:
add missing iscsi pod definition

Signed-off-by: Huamin Chen <hchen@redhat.com>

rootfs commented:
moved to a cleaner PR #5506 
@thockin @eparis @markturansky

thockin commented:
Will review ASAP

On Mon, Mar 16, 2015 at 7:26 AM, Huamin Chen <notifications@github.com>
wrote:

> Closed #4612 <https://github.com/GoogleCloudPlatform/kubernetes/pull/4612>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4612#event-255566377>
> .
>
